### PR TITLE
add local_ratelimit descriptor example

### PR DESCRIPTION
**Pull request by StarryVae**
Signed-off-by: wangkai19 <wangkai19@corp.netease.com>

Commit Message: add local_ratelimit descriptor example
Additional Description: N/A
Risk Level: None
Testing: N/A
Docs Changes: N/A
Release Notes: N/A
Platform Specific Features: N/A

### CODE REVIEW

**Review by phlax — COMMENTED** [review]
hi @StarryVae thanks for working on this

if we update the example, then i think we also need to update the example docs to demonstrate what is being changed here

**Comment by StarryVae** [comment]
> hi @StarryVae thanks for working on this
> 
> if we update the example, then i think we also need to update the example docs to demonstrate what is being changed here

@phlax Example doc has been updated, PTAL again, thanks.

File: examples/local_ratelimit/ratelimit-envoy.yaml [hunk 1]
```diff
@@ -95,6 +99,14 @@ static_resources:
                     key: x-local-rate-limit
                     value: 'true'
               local_rate_limit_per_downstream_connection: false
+              descriptors:
+              - entries:
```
Comment (daixiang0) [inline examples/local_ratelimit/ratelimit-envoy.yaml:103]: If we want to show how descriptors work, two descriptors with different token bucket settings make more sense.

Comment (StarryVae) [reply examples/local_ratelimit/ratelimit-envoy.yaml:103]: thanks, PTAL again.

**Comment by phlax** [comment]
docs are rendered here https://storage.googleapis.com/envoy-pr/20868/docs/index.html

**Review by phlax — COMMENTED** [review]
@StarryVae i still dont see anything about descriptors mentioned in the docs for this example

if we are going to add more configuration to the example it should exemplify something and have documentation as to what effect it has

without this, the example would be better being as simple as possible

maybe im missing something tho - is this just fixing something?

**Comment by daixiang0** [comment]
@phlax seems those commits are about flaky tests. Where can I check it is flaky or not recently?

**Comment by StarryVae** [comment]
> @StarryVae i still dont see anything about descriptors mentioned in the docs for this example
> 
> if we are going to add more configuration to the example it should exemplify something and have documentation as to what effect it has
> 
> without this, the example would be better being as simple as possible
> 
> maybe im missing something tho - is this just fixing something?

@phlax Sorry, i have added descriptors to the doc, PTAL again, thank you.

**Comment by StarryVae** [comment]
> @phlax seems those commits are about flaky tests. Where can I check it is flaky or not recently?
> 
> @StarryVae please update docs refer to above comment.

@daixiang0 docs are updated, PTAL again, thank you.

File: docs/root/start/sandboxes/local_ratelimit.rst [hunk 2]
```diff
@@ -41,7 +41,9 @@ Step 2: Test rate limiting of upstream service
 
 The sandbox is configured with ``10000`` port for upstream service.
 
-If a request reaches the rate limit, Envoy will add ``x-local-rate-limit`` header and refuse the connection with a 429 HTTP response code and with the content ``local_rate_limited``.
+Two descriptors with generic_key are set in this example, and the route will match the descriptor with generic_key ``test1``.
+
+If a request reaches the descriptor ``test1's`` rate limit, Envoy will add ``x-local-rate-limit`` header and refuse the connection with a 429 HTTP response code and with the content ``local_rate_limited``.
```
Comment (daixiang0) [inline docs/root/start/sandboxes/local_ratelimit.rst:46]: If a request reaches the rate limit of descriptor ``test1`` ...

Comment (StarryVae) [reply docs/root/start/sandboxes/local_ratelimit.rst]: thanks ...

**Review by wbpcode — APPROVED** [review]
LGTM. Thanks. @StarryVae

**Comment by phlax** [comment]
docs are rendered here https://storage.googleapis.com/envoy-pr/20868/docs/index.html

File: docs/root/start/sandboxes/local_ratelimit.rst [hunk 3]
```diff
@@ -41,7 +41,9 @@ Step 2: Test rate limiting of upstream service
 
 The sandbox is configured with ``10000`` port for upstream service.
 
-If a request reaches the rate limit, Envoy will add ``x-local-rate-limit`` header and refuse the connection with a 429 HTTP response code and with the content ``local_rate_limited``.
+Two descriptors with generic_key are set in this example, and the route will match the descriptor with generic_key ``test1``.
```
Comment (phlax) [inline docs/root/start/sandboxes/local_ratelimit.rst:44]:
```suggestion
Two descriptors with ``generic_key`` are set in this example, and the route will match the descriptor with ``generic_key `` ``test1``.
```

File: examples/local_ratelimit/ratelimit-envoy.yaml [hunk 4]
```diff
@@ -95,6 +99,21 @@ static_resources:
...
+                - key: generic_key
+                  value: test1
+                token_bucket:
+                  max_tokens: 3
+                  tokens_per_fill: 3
+                  fill_interval: 6s
+              - entries:
```
Comment (phlax) [inline examples/local_ratelimit/ratelimit-envoy.yaml:110]: this doesnt appear to be used in either the docs or the verification - is it necessary ?

File: examples/local_ratelimit/ratelimit-envoy.yaml [hunk 5]
```diff
@@ -78,7 +82,7 @@ static_resources:
               token_bucket:
                 max_tokens: 2
                 tokens_per_fill: 2
-                fill_interval: 5s
+                fill_interval: 2s
```
Comment (phlax) [inline examples/local_ratelimit/ratelimit-envoy.yaml:85]: is this correct ? i think it was changed previously to resolve a flake

**Review by phlax — CHANGES_REQUESTED** [review]
@StarryVae thanks again for working on this

i have left a couple of comments inline

overall reading through the docs im not quite getting the significance of the descriptors

given the example worked previously without any descriptors, then im minded to keep an example that doesnt use them and separate out a next step that demonstrates the difference

its not immediately clear to me how the descriptors work in terms of matching and so on - im guessing you can use a descriptor across multiple listeners/routes and they share rate limiting rules and bucket - i think we need to make this clearer

relatedly, i then am wondering - what are the defaults - ie if some routes are rate-limited with a descriptor what happens with a route configured to rate=limit but with no descriptor set

File: examples/local_ratelimit/ratelimit-envoy.yaml (hunk 4 above)
Comment (StarryVae) [reply examples/local_ratelimit/ratelimit-envoy.yaml:110]: it is the specific rate limit of descriptor `test1`, the route of upstream service in listener 10000 will match descriptor `test1`, as a result, this route can be requested 3 times at most in 6s, at the same time, the default rate limit will not be applied.

File: examples/local_ratelimit/ratelimit-envoy.yaml (hunk 5 above)
Comment (StarryVae) [reply examples/local_ratelimit/ratelimit-envoy.yaml:85]: the default rate limit will not be applied except the default `fill_interval` when descriptors are set. It means that the local rate_limit plugin will check if the buckets needs to be filled every default `fill_interval`, so in this example,  the `fill_interval` of descriptors is set 6s, the default `fill_interval` should be set to the factor of 6s such as 2s.

File: docs/root/start/sandboxes/local_ratelimit.rst (hunk 3 above)
Comment (StarryVae) [reply docs/root/start/sandboxes/local_ratelimit.rst]: done, thanks.

**Comment by StarryVae** [comment]
@phlax PTAL again, thanks.

**Comment by alyssawilk** [comment]
@phlax ping!

**Review by phlax — COMMENTED** [review]
@StarryVae looking at the docs here https://storage.googleapis.com/envoy-pr/793be16/docs/start/sandboxes/local_ratelimit.html its not clear to me at all from reading that how the descriptors work

also, as i explained previously - the rate limiting works _without_ using descriptors then i want to keep that example, and separate out the example with descriptors

the important thing to remember with these sandboxes is that they are intended to convey a basic understanding of what is being shown to someone with ~zero knowledge

**Comment by phlax** [comment]
i would like to add information about descriptors to this example, so im trying to think of a way forward

here is my suggestions:

- update the example config so there are routes that work both with and without descriptors
- add a new section "Rate limiting with descriptors" or similar and go through the steps for the descriptor config there
- add a brief explanation about the descriptor *in the introduction parargraph"

**Comment by StarryVae** [comment]
> i would like to add information about descriptors to this example, so im trying to think of a way forward
> 
> here is my suggestions:
> 
> * update the example config so there are routes that work both with and without descriptors
> * add a new section "Rate limiting with descriptors" or similar and go through the steps for the descriptor config there
> * add a brief explanation about the descriptor *in the introduction parargraph"

Thanks, i will try to fix these suggestions.

**Comment by daixiang0** [comment]
friendly ping @StarryVae

**Comment by StarryVae** [comment]
Sorry for late, i will try complete it as soon as possible.

**Pull request closed without merging**